src/functions.c

```diff
@@ -189,18 +189,22 @@ void functionsLibCtxClearCurrent(int async) {
 }
 
 /* Free the given functions ctx */
-void functionsLibCtxFree(functionsLibCtx *functions_lib_ctx) {
-    functionsLibCtxClear(functions_lib_ctx);
-    dictRelease(functions_lib_ctx->functions);
-    dictRelease(functions_lib_ctx->libraries);
-    dictRelease(functions_lib_ctx->engines_stats);
-    zfree(functions_lib_ctx);
+void functionsLibCtxFree(functionsLibCtx *functions_lib_ctx, int async) {
+    if (async) {
+        freeFunctionsAsync(functions_lib_ctx);
+    } else {
+        functionsLibCtxClear(functions_lib_ctx);
+        dictRelease(functions_lib_ctx->functions);
+        dictRelease(functions_lib_ctx->libraries);
+        dictRelease(functions_lib_ctx->engines_stats);
+        zfree(functions_lib_ctx);
+    }
```

Comment:
freeFunctionsAsync calls functionsLibCtxFree again, so we get a circular call graph.

```
    functionsLibCtxFree --> freeFunctionsAsync
           ^                        /
            \                      /
             `--------------------'
```

Even if freeFunctionsAsync always calls it with async=0, it's not a perfect design IMO. I think it's better something like this:

```
            functionsLibCtxFreeSyncOrAsync
                   /           \
                  /             \
                 v               v
    functionsLibCtxFree <-- freeFunctionsAsync
```

Here, functionsLibCtxFree is only sync, as before. If we do this, then functionsLibCtxFreeSyncOrAsync can be static and internal in functions.c.

Reply:
yean, that is a bad call, i added a new freeFunctionsSync just like scriptingRelease, does it better?

Reply:
It's better but I was also suggesting that we don't need both of them in functions.h. All the calls from files outside functions.c only need to call sync function.

Also, to minimize the diff, I suggested to keep functionsLibCtxFree as the sync function, so existing calls to this one can remain identical. We can add a new one, which is internal in functions.c, that has the async parameter.

Do you think it's better that the sync function has "sync" in the name?

Reply:
sorry, it was so long ago that i forgot the origin diff and only looked at the most recent one. 

> Do you think it's better that the sync function has "sync" in the name?

i don't mind actually... scriptingRelease has the same style (this is actully written by me...)

```
void scriptingRelease(int async) {
    if (async)
        freeLuaScriptsAsync(lctx.lua_scripts, lctx.lua_scripts_lru_list, lctx.lua);
    else
        freeLuaScriptsSync(lctx.lua_scripts, lctx.lua_scripts_lru_list, lctx.lua);
}
```